jaxrs-api/src/test/java/javax/ws/rs/core/MediaTypeTest.java

```diff
@@ -58,4 +60,110 @@ public void testNullConstructorValues() {
         actual = new MediaType(null, null, (String) null);
         assertEquals(MediaType.WILDCARD_TYPE, actual);
     }
+
+    /**
+     * Test that a wildcard type {@link MediaType} value with explicit subtype is not compatible with
+     * a non-wildcard type {@link MediaType} value with a different subtype.
+     */
+    @Test
+    public void testMediaTypeWithWildcardTypeNotCompatibleWhenSubtypeDifferent() {
+        MediaType wildcard = new MediaType(MediaType.MEDIA_TYPE_WILDCARD, "json");
+
+        assertFalse(wildcard.isCompatible(MediaType.APPLICATION_OCTET_STREAM_TYPE));
+        assertFalse(MediaType.APPLICATION_OCTET_STREAM_TYPE.isCompatible(wildcard));
+    }
+
+    /**
+     * Test that a wildcard type {@link MediaType} value with explicit subtype is compatible with
+     * a non-wildcard type {@link MediaType} value with the same explicit subtype.
+     */
+    @Test
+    public void testMediaTypeWithWildcardTypeCompatibleWhenSubtypeMatches() {
+        MediaType wildcard = new MediaType(MediaType.MEDIA_TYPE_WILDCARD, "json");
+
+        assertTrue(wildcard.isCompatible(MediaType.APPLICATION_JSON_TYPE));
+        assertTrue(MediaType.APPLICATION_JSON_TYPE.isCompatible(wildcard));
+    }
+
+    /**
+     * Test that a wildcard type {@link MediaType} value with explicit subtype is not compatible with
+     * a non-wildcard type {@link MediaType} value with a different subtype.
+     */
+    @Test
+    public void testMediaTypeWithWildcardTypeCompatibleWithExplicitTypeAndWildcardSubtype() {
+        MediaType wildcardType = new MediaType(MediaType.MEDIA_TYPE_WILDCARD, "json");
+        MediaType wildcardSubtype = new MediaType("application", MediaType.MEDIA_TYPE_WILDCARD);
+
+        assertTrue(wildcardType.isCompatible(wildcardSubtype));
+        assertTrue(wildcardSubtype.isCompatible(wildcardType));
+    }
+
+    /**
+     * Test that a {@link MediaType} value with explicit type and wildcard subtype
+     * is compatible with a {@link MediaType} value with the same type and an explicit subtype.
+     */
+    @Test
+    public void testMediaTypeWithWildcardSubtypeCompatibleWhenTypeMatches() {
+        MediaType wildcard = new MediaType("application", MediaType.MEDIA_TYPE_WILDCARD);
+
+        assertTrue(wildcard.isCompatible(MediaType.APPLICATION_JSON_TYPE));
```

Comment:
Actually a more specific naming would make it easier to understand the tests, so the comments would be mostly unnecessary:
```java
assertThat(mediaTypeApplicationAny, isCompatibleWith(MediaType.APPLICATION_JSON_TYPE));
```